src/data/matrix/kronecker.lean

```diff
@@ -0,0 +1,317 @@
+  kronecker_map f Aᵀ Bᵀ = (kronecker_map f A B)ᵀ :=
+ext $ λ i j, rfl
+
+lemma kronecker_map_map_left (f : α' → β → γ) (g : α → α')
+  (A : matrix l m α) (B : matrix n p β) :
+  kronecker_map f (A.map g) B = kronecker_map (λ a b, f (g a) b) A B :=
+ext $ λ i j, rfl
+
+lemma kronecker_map_map_right (f : α → β' → γ) (g : β → β')
+  (A : matrix l m α) (B : matrix n p β) :
+  kronecker_map f A (B.map g) = kronecker_map (λ a b, f a (g b)) A B :=
+ext $ λ i j, rfl
+
+lemma kronecker_map_map (f : α → β → γ) (g : γ → γ')
+  (A : matrix l m α) (B : matrix n p β) :
+  (kronecker_map f A B).map g = kronecker_map (λ a b, g (f a b)) A B :=
+ext $ λ i j, rfl
+
+lemma kronecker_map_zero_left [has_zero α] [has_zero γ] (f : α → β → γ) (hf : ∀ b, f 0 b = 0)
+  (B : matrix n p β) :
+  kronecker_map f (0 : matrix l m α) B = 0:=
+ext $ λ i j,hf _
+
+lemma kronecker_map_zero_right [has_zero β] [has_zero γ] (f : α → β → γ) (hf : ∀ a, f a 0 = 0)
+  (A : matrix l m α) :
+  kronecker_map f A (0 : matrix n p β) = 0 :=
+ext $ λ i j, hf _
+
+lemma kronecker_map_add_left [has_add α] [has_add γ] (f : α → β → γ)
+  (hf : ∀ a₁ a₂ b, f (a₁ + a₂) b = f a₁ b + f a₂ b)
+  (A₁ A₂ : matrix l m α) (B : matrix n p β) :
+  kronecker_map f (A₁ + A₂) B = kronecker_map f A₁ B + kronecker_map f A₂ B :=
+ext $ λ i j, hf _ _ _
+
+lemma kronecker_map_add_right [has_add β] [has_add γ] (f : α → β → γ)
+  (hf : ∀ a b₁ b₂, f a (b₁ + b₂) = f a b₁ + f a b₂)
+  (A : matrix l m α) (B₁ B₂ : matrix n p β) :
+  kronecker_map f A (B₁ + B₂) = kronecker_map f A B₁ + kronecker_map f A B₂ :=
+ext $ λ i j, hf _ _ _
+
+lemma kronecker_map_smul_left [has_scalar R α] [has_scalar R γ] (f : α → β → γ)
+  (hf : ∀ (r : R) a b, f (r • a) b = r • f a b) (r : R)
+  (A : matrix l m α) (B : matrix n p β) :
+  kronecker_map f (r • A) B = r • kronecker_map f A B :=
+ext $ λ i j, hf _ _ _
+
+lemma kronecker_map_smul_right [has_scalar R β] [has_scalar R γ] (f : α → β → γ)
+  (hf : ∀ (r : R) a b, f a (r • b) = r • f a b) (r : R)
+  (A : matrix l m α) (B : matrix n p β) :
+  kronecker_map f A (r • B) = r • kronecker_map f A B :=
+ext $ λ i j, hf _ _ _
+
+lemma kronecker_map_diagonal_diagonal [has_zero α] [has_zero β] [has_zero γ]
+  [decidable_eq m] [decidable_eq n]
+  (f : α → β → γ) (hf₁ : ∀ b, f 0 b = 0) (hf₂ : ∀ a, f a 0 = 0) (a : m → α) (b : n → β):
+  kronecker_map f (diagonal a) (diagonal b) = diagonal (λ mn, f (a mn.1) (b mn.2)) :=
+begin
+  ext ⟨i₁, i₂⟩ ⟨j₁, j₂⟩,
+  simp [diagonal, apply_ite f, ite_and, ite_apply, apply_ite (f (a i₁)), hf₁, hf₂],
+end
+
+lemma kronecker_map_one_one [has_zero α] [has_zero β] [has_zero γ]
+  [has_one α] [has_one β] [has_one γ]
+  [decidable_eq m] [decidable_eq n]
+  (f : α → β → γ) (hf₁ : ∀ b, f 0 b = 0) (hf₂ : ∀ a, f a 0 = 0) (hf₃ : f 1 1 = 1) :
+  kronecker_map f (1 : matrix m m α) (1 : matrix n n β) = 1 :=
+(kronecker_map_diagonal_diagonal _ hf₁ hf₂ _ _).trans $ by simp only [hf₃, diagonal_one]
+
+/-- When `f` is bilinear then `matrix.kronecker_map f` is also bilinear. -/
```

Comment:
Why `kronecker_map_linear` rather than `kronecker_map_bilinear`?

Reply:
I don't have a particularly good reason; I chose `linear` because the type is `linear_map`. I'm happy to change it.

Reply:
Thanks!

Reply:
Renamed